### PR TITLE
Separate setting build environment and run environment in packages

### LINKED ISSUE #2016: setup_environment() is not mentioned in Packaging Guide
subj. http://spack.readthedocs.io/en/latest/packaging_guide.html

### LINKED ISSUE #6143: Error building elpa: AttributeError: 'Spec' object has no attribute 'mpicc'
I was trying to do: `spack install elpa %gcc@6.4.0 ^intel-mkl@2018.0.128 ^intel-mpi@2018.0.128` on Rocks cluster 6.2/CentOS 6.6, but in the install phase, it gives the error in the title. The spack-build.out indicates however that the build was successful:

    ----------------------------------------------------------------------
    Libraries have been installed in:
       /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/lib

    If you ever happen to want to link against installed libraries
    in a given directory, LIBDIR, you must either use libtool, and
    specify the full pathname of the library, or use the '-LLIBDIR'
    flag during linking and do at least one of the following:
      - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
        during execution
      - add LIBDIR to the 'LD_RUN_PATH' environment variable
        during linking
      - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
      - have your system administrator add LIBDIR to '/etc/ld.so.conf'
    
    See any operating system documentation about shared libraries for
    more information, such as the ld(1) and ld.so(8) manual pages.
     ----------------------------------------------------------------------
    /bin/mkdir -p '/share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/bin'
     /bin/sh ./libtool   --mode=install /usr/bin/install -c elpa2_print_kernels '/share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/bin'
    libtool: install: /usr/bin/install -c .libs/elpa2_print_kernels /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/bin/elpa2_print_kernels
    make[2]: Leaving directory `/tmp/soft/spack-stage/spack-stage-aaRYV2/elpa-2016.05.004/spack-build'
    make[1]: Leaving directory `/tmp/soft/spack-stage/spack-stage-aaRYV2/elpa-2016.05.004/spack-build'

The full error logs are:

    ==> intel-mkl is already installed in /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/intel-mkl-2018.0.128-ghtmbtscrmcextiiwijcgh2pv2xftq6r                                       
    ==> intel-mpi is already installed in /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/intel-mpi-2018.0.128-hfjbyb3cljw5ej2xtil6vujvjrvywbc2                                       
    ==> Installing elpa                                                                                                                                                                          
    ('spec-mpi', ['__class__', '__contains__', '__delattr__', '__dict__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_add_dependency', '_add_flag', '_add_version', '_autospec', '_cmp_key', '_cmp_key_cache', '_cmp_node', '_concrete', '_concretize_helper', '_constrain_dependencies', '_dependencies', '_dependents', '_dup', '_dup_deps', '_eq_dag', '_evaluate_dependency_conditions', '_expand_virtual_packages', '_find_deps', '_find_provider', '_hash', '_install_status', '_installed_explicitly', '_mark_concrete', '_merge_dependency', '_normal', '_normalize_helper', '_replace_with', '_set_architecture', '_set_compiler', '_set_test_prefix', 'architecture', 'cformat', 'colorized', 'command', 'common_dependencies', 'compiler', 'compiler_flags', 'concrete', 'concretize', 'concretized', 'constrain', 'constrained', 'copy', 'cshort_spec', 'dag_hash', 'dag_hash_bit_prefix', 'dep_difference', 'dep_string', 'dependencies', 'dependencies_dict', 'dependents', 'dependents_dict', 'eq_dag', 'eq_node', 'external', 'external_module', 'external_path', 'flat_dependencies', 'format', 'from_dict', 'from_json', 'from_literal', 'from_node_dict', 'from_yaml', 'fullname', 'get_dependency', 'headers', 'index', 'is_virtual', 'last_query', 'libs', 'mpicc', 'mpicxx', 'mpif77', 'mpifc', 'name', 'namespace', 'ne_dag', 'ne_node', 'normalize', 'normalized', 'package', 'package_class', 'patches', 'prefix', 'read_yaml_dep_specs', 'root', 'satisfies', 'satisfies_dependencies', 'short_spec', 'sorted_deps', 'to_dict', 'to_json', 'to_node_dict', 'to_yaml', 'traverse', 'traverse_edges', 'tree', 'validate_or_raise', 'variants', 'version', 'versions', 'virtual', 'virtual_dependencies'])
    ==> Warning: Suspicious requests to set or unset 'CC' found
    ==> Warning:            env.set('CC', join_path(link_dir, compiler.link_paths['cc'])) at /share/apps/spack/lib/spack/spack/build_environment.py:148
    ==> Warning:    --->    spack_env.set('CC', spec['mpi'].mpicc) at /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:72
    ==> Warning: Suspicious requests to set or unset 'CXX' found
    ==> Warning:            env.set('CXX', join_path(link_dir, compiler.link_paths['cxx'])) at /share/apps/spack/lib/spack/spack/build_environment.py:151
    ==> Warning:    --->    spack_env.set('CXX', spec['mpi'].mpicxx) at /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:74
    ==> Warning: Suspicious requests to set or unset 'FC' found
    ==> Warning:            env.set('FC', join_path(link_dir, compiler.link_paths['fc'])) at /share/apps/spack/lib/spack/spack/build_environment.py:157
    ==> Warning:    --->    spack_env.set('FC', spec['mpi'].mpifc) at /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:73
    ==> Using cached archive: /share/apps/spack/var/spack/cache/elpa/elpa-2016.05.004.tar.gz
    ==> Staging archive: /share/apps/spack/var/spack/stage/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/elpa-2016.05.004.tar.gz
    ==> Created stage in /share/apps/spack/var/spack/stage/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj
    ==> No patches needed for elpa
    ==> Building elpa [AutotoolsPackage]
    ==> Executing phase: 'autoreconf'
    ==> Executing phase: 'configure'
    ==> Executing phase: 'build'
    ==> Executing phase: 'install'
    ('spec-mpi', ['__class__', '__contains__', '__delattr__', '__dict__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_add_dependency', '_add_flag', '_add_version', '_autospec', '_cmp_key', '_cmp_key_cache', '_cmp_node', '_concrete', '_concretize_helper', '_constrain_dependencies', '_dependencies', '_dependents', '_dup', '_dup_deps', '_eq_dag', '_evaluate_dependency_conditions', '_expand_virtual_packages', '_find_deps', '_find_provider', '_hash', '_install_status', '_installed_explicitly', '_mark_concrete', '_merge_dependency', '_normal', '_normalize_helper', '_replace_with', '_set_architecture', '_set_compiler', '_set_test_prefix', 'architecture', 'cformat', 'colorized', 'command', 'common_dependencies', 'compiler', 'compiler_flags', 'concrete', 'concretize', 'concretized', 'constrain', 'constrained', 'copy', 'cshort_spec', 'dag_hash', 'dag_hash_bit_prefix', 'dep_difference', 'dep_string', 'dependencies', 'dependencies_dict', 'dependents', 'dependents_dict', 'eq_dag', 'eq_node', 'external', 'external_module', 'external_path', 'flat_dependencies', 'format', 'from_dict', 'from_json', 'from_literal', 'from_node_dict', 'from_yaml', 'fullname', 'get_dependency', 'headers', 'index', 'is_virtual', 'last_query', 'libs', 'name', 'namespace', 'ne_dag', 'ne_node', 'normalize', 'normalized', 'package', 'package_class', 'patches', 'prefix', 'read_yaml_dep_specs', 'root', 'satisfies', 'satisfies_dependencies', 'short_spec', 'sorted_deps', 'to_dict', 'to_json', 'to_node_dict', 'to_yaml', 'traverse', 'traverse_edges', 'tree', 'validate_or_raise', 'variants', 'version', 'versions', 'virtual', 'virtual_dependencies'])
    ==> Error: AttributeError: 'Spec' object has no attribute 'mpicc'

    /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:71, in setup_environment:
         3             spec = self.spec
         4
         5             print('spec-mpi', dir(spec['mpi']))
      >> 6             spack_env.set('CC', spec['mpi'].mpicc)
         7             spack_env.set('FC', spec['mpi'].mpifc)
         8             spack_env.set('CXX', spec['mpi'].mpicxx)

### PR DESCRIPTION
**Pull request by alalazo**
fixes #6143
fixes #2016

refers to https://github.com/spack/spack/issues/2016#issuecomment-258884636

As a package maintainer I want different methods to specify the build-time environment and the run-time environment of a package, so that I will know exactly which are the preconditions of the method that is being called.

### Rationale

Currently each package can override two methods:
```python
def setup_environment(self, build_env, run_env):
    ...

def setup_dependent_environment(self, build_env, run_env, dependent_spec):
    ...
```
to modify its build or run environment, or those of a dependent. This has a fundamental issue in that the same method will be called at different times, and will have different preconditions depending when it's called:

1. If it is called to set-up the build environment the prefix of the spec being installed is not yet there + a few special variables that are needed to build with Spack compiler wrappers will be set
2. If it is called to set-up the run environment the underlying spec is assumed to be already installed

This in turn might cause issues like #6143 or https://github.com/spack/spack/issues/2016#issuecomment-258884636. In this PR we split each of the two methods above, e.g. :
```python
def setup_build_environment(self, env):
   ...

def setup_run_environment(self, env):
   ...
```
so that it will be clear which are the preconditions of each method call.

### Description

This PR splits each `*_environment` method in `spack.package.PackageBase` into two: one that will be called for the build-time environment and one that wil be called for the run-time. As there are a lot of packages overriding these callbacks the default implementation is such to mimic the legacy behavior, and emit a deprecation warning to `tty.debug`, e.g.:
```console
$ spack -d install mpich
[...]
==> [2019-04-05-14:13:38.193971] Installing mpich
==> [2019-04-05-14:13:38.194114] Searching for binary cache of mpich
==> [2019-04-05-14:13:38.194407] Reading config file /home/mculpo/.spack/linux/mirrors.yaml
==> [2019-04-05-14:13:38.195858] Finding buildcaches in /home/mculpo/production/mirror/build_cache
==> [2019-04-05-14:13:38.198158] No binary for mpich found: installing from source
==> [2019-04-05-14:13:38.300407] [DEPRECATED METHOD]
"pkgconf" still defines the deprecated method "setup_dependent_environment" [should be split into "setup_dependent_build_environment" and "setup_dependent_run_environment"]
==> [2019-04-05-14:13:38.323784] [DEPRECATED METHOD]
"perl" still defines the deprecated method "setup_dependent_environment" [should be split into "setup_dependent_build_environment" and "setup_dependent_run_environment"]
==> [2019-04-05-14:13:38.344990] [DEPRECATED METHOD]
"libtool" still defines the deprecated method "setup_dependent_environment" [should be split into "setup_dependent_build_environment" and "setup_dependent_run_environment"]
[...]
```

### Additional information

- [x] Split `setup_environment` into two methods
- [x] Split `setup_dependent_environment` into two methods
- [x] Add documentation in the "Packaging guide" section of the docs
- [x] Update the tutorial sections of the docs
- [x] Add a test that fails if a package under PR still implements the old methods (this can be used to make the community aware of the new methods, can be removed once the old methods are not supported anymore)
- [x] Add a test that fails if any package implements the old methods, and mark it `xfail` (it will be useful when all packages are ported, in case of people doing PR based on old behavior)

### CODE REVIEW

**Comment by alalazo** [comment]
I added this PR as a draft to collect opinions on splitting the two methods before proceeding with docs and tests. If you don't have major concerns on what's being done here we can use the PR also to improve the APIs in `build_environment.py` and serve cases like #11084. Let me know what you think.

File: lib/spack/spack/build_environment.py [hunk 1]
```diff
@@ -665,35 +665,26 @@ def load_external_modules(pkg):
...
+    pkg.architecture.platform.setup_platform_environment(pkg, build_env)
 
-    # traverse in postorder so package can use vars from its dependencies
-    spec = pkg.spec
-    for dspec in pkg.spec.traverse(order='post', root=False,
-                                   deptype=('build', 'test')):
-        spkg = dspec.package
-        set_module_variables_for_package(spkg)
+    build_env.extend(
+        environment_modifications_from_dependencies(pkg.spec, context='build')
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]: I think this method name should be changed. It doesn't only modify the environment, it also modifies the python module scope.

**Comment by tgamblin** [comment]
No deep review yet, but this is a much cleaner, less confusing direction than the current state of affairs.

File: lib/spack/spack/build_environment.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/build_environment.py]: Done

**Comment by alalazo** [comment]
Ready for a review from my side. The description at the top should be up to date. Unit tests have been added and docs updated.

File: lib/spack/spack/build_environment.py [hunk 2]
```diff
@@ -722,8 +713,38 @@ def setup_package(pkg, dirty):
...
+        spec (Spec): spec for which we want the modifications
+        context (str): either 'build' for build-time modifications or 'run'
+            for run-time modifications
+    """
+    env = EnvironmentModifications()
+    pkg = spec.package
+
+    # Maps the context to deptype and method to be called
+    deptype_and_method = {
+        'build': (('build', 'test'), 'setup_dependent_build_environment'),
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]: I think link dependencies should call `setup_dependent_build_environment` if their dependent is building. I guess this goes back to the conversation about what `link` vs `('build', 'link')` dependencies mean.

Comment (becker33) [inline lib/spack/spack/build_environment.py]: I think link dependencies should call setup_dependent_build_environment if their dependent is building. I guess this goes back to the conversation about what link vs ('build', 'link') dependencies mean.

Comment (alalazo) [reply lib/spack/spack/build_environment.py]: It makes a lot of sense, since `link` are also responsible of all the include flags right now.

Comment (alalazo) [reply lib/spack/spack/build_environment.py]: Should be fixed by c65bfc9

File: lib/spack/spack/cmd/flake8.py [hunk 3]
```diff
@@ -246,7 +245,7 @@ def prefix_relative(path):
 
         with working_dir(spack.paths.prefix):
             if not file_list:
-                file_list = changed_files(args)
+                file_list = changed_files(args.base, args.untracked, args.all)
```
Comment (becker33) [inline lib/spack/spack/cmd/flake8.py]: Why are there changes to `spack flake8` in this PR? I don't see why these are related.

**Comment by alalazo** [comment]
@becker33 

> Are there other packages in builtin, builtin.mock, or tutorial repos that aren't updated in the PR?

There are a lot of packages that are not updated in this PR. More details in the description above (which is up to date despite being posted a long time ago), but the bottomline is that changes were meant to be incremental and the default implementation for the new methods falls back to the old one and prints a warning in the debug log.

To avoid that people continue using the old methods there are 2 unit tests added:

- One is marked `xfail` and is meant to be unmarked once the conversion of all packages is done in the future. It checks that we have no packages which implement deprecated methods.
- The other one is temporary, and checks that for any package changed in a PR that package has the `setup_*` methods updated too.

> Why are there changes to spack flake8 in this PR? I don't see why these are related.

These changes were done to implement the second unit test above and reuse most of the functionality to retrieve the files changed by any given PR. See https://github.com/spack/spack/pull/11115/commits/d28363090c6f3af24df78a265c2581c829c71fff

File: lib/spack/spack/cmd/flake8.py (hunk 3 above)
Comment (alalazo) [reply lib/spack/spack/cmd/flake8.py]: Replied below

**Pull request merged**